### PR TITLE
client: better error handling

### PR DESCRIPTION
**Pull request by tbillington**

### CODE REVIEW

File: src/sajari.js [hunk 1]
```diff
@@ -23,13 +23,32 @@ const makeRequest = (address, body, callback) => {
...
-    } else {
+      try {
+        callback(null, JSON.parse(request.responseText));
+      } catch (e) {
+        callback("Error parsing response.", null);
+      }
+      return;
+    }
+
+    if (request.status === 403) {
```
Comment (dhowden) [inline src/sajari.js]: Also make sure that the actual error is kept around too (i.e. so that you can log it)?

File: src/sajari.js [hunk 2]
```diff
@@ -23,13 +23,32 @@ const makeRequest = (address, body, callback) => {
     if (request.readyState !== 4) return;
 
     if (request.status === 200) {
-      callback(null, JSON.parse(request.responseText));
-    } else {
+      try {
+        callback(null, JSON.parse(request.responseText));
```
Comment (dhowden) [inline src/sajari.js]: Doesn't this try-catch also apply to errors thrown from within the callback?  If you're wanting to catch errors from JSON.parse it should probably be seperate?

File: src/sajari.js [hunk 3]
```diff
@@ -23,13 +23,32 @@ const makeRequest = (address, body, callback) => {
...
-        request.response || request.responseText || "error connecting",
+        "Authorisation for this request failed. Check your credentials.",
         null
       );
+      return;
     }
+
+    if (request.responseText) {
+      try {
+        callback(JSON.parse(request.responseText).message, null);
```
Comment (dhowden) [inline src/sajari.js]: Ditto try-catch comment above with regards to errors thrown from `callback`

File: src/sajari.js [hunk 4]
```diff
@@ -22,33 +22,21 @@ const makeRequest = (address, body, callback) => {
...
-      } catch (e) {
-        callback("Error parsing response.", null);
-      }
-      return;
+    let parsedResponse;
+    try {
+      parsedResponse = JSON.parse(request.responseText);
+    } catch (e) {
+      const error = { message: "Error parsing response" };
+      callback(error, null);
```
Comment (dhowden) [inline src/sajari.js]: You need to return after this?

Comment (tbillington) [reply src/sajari.js]: Was getting there :P !

**Pull request merged**